provider/plugin/plugin.go

```diff
@@ -0,0 +1,337 @@
+	err = backoff.Retry(func() error {
+		resp, err = client.Do(req)
+		if err != nil {
+			log.Printf("error connecting to plugin api: %v", err)
+		}
+		return err
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetries))
+
+	vary := resp.Header.Get(varyHeader)
+	contentType := resp.Header.Get(contentTypeHeader)
+
+	if vary != contentTypeHeader {
+		return nil, fmt.Errorf("wrong vary value returned from server: %s", vary)
+	}
+
+	if contentType != mediaTypeFormatAndVersion {
+		return nil, fmt.Errorf("wrong content type returned from server: %s", contentType)
+	}
+
+	return &PluginProvider{
+		client:          client,
+		remoteServerURL: parsedURL,
+	}, nil
+}
+
+// Records will make a GET call to remoteServerURL/records and return the results
+func (p PluginProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	u, err := url.JoinPath(p.remoteServerURL.String(), "records")
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("error joining path: %s", err.Error())
+		return nil, err
+	}
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("error creating request: %s", err.Error())
+		return nil, err
+	}
+	resp, err := p.client.Do(req)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("error performing request: %s", err.Error())
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		recordsErrorsGauge.Inc()
+		log.Debugf("error from external provider, HTTP status code %d", resp.StatusCode)
+		return nil, fmt.Errorf("failed to apply changes with code %d", resp.StatusCode)
+	}
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("error reading response body: %s", err.Error())
+		return nil, err
+	}
+
+	endpoints := []*endpoint.Endpoint{}
+	err = json.Unmarshal(b, &endpoints)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("error unmarshaling response body: %s", err.Error())
+		return nil, err
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges will make a POST to remoteServerURL/records with the changes
+func (p PluginProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	u, err := url.JoinPath(p.remoteServerURL.String(), "records")
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("error joining path: %s", err.Error())
+		return err
+	}
+	b, err := json.Marshal(changes)
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("error marshaling changes: %s", err.Error())
+		return err
+	}
+
+	req, err := http.NewRequest("POST", u, bytes.NewBuffer(b))
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("error creating request: %s", err.Error())
+		return err
+	}
+	resp, err := p.client.Do(req)
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("error performing request: %s", err.Error())
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
```

Comment:
I think the docs say StatusCode 20x is fine, but here it's only 200, so we should either change docs or what we expect here.